---
title: "T1"
author: "Atom"
date: "2/6/2022"
output: html_document
---

```{r}
count_gene_mutations <- function (table, gene.col, gt.col, flag = c("Alleles", "Snps", "Carrier")) {
  
  if (is.double(gene.col) == FALSE){
    return ("gene.col needs to be an integer.")
  }
  
  if (is.double(gt.col) == FALSE){
    return ("gt.col needs to be an integer.")
  }
  
  flag <- match.arg(flag)
 
  newTable <- dplyr::select(table, c(gt.col:ncol(table)))
  
  gt2count <- function(genotype){
    stringr::str_extract_all(genotype, "\\d") %>%
    purrr::map_int(~sum(as.integer(.)))
  }
  
  topTable <- newTable %>%
    purrr::map_df(gt2count)
  
  BottomTable <- topTable
  BottomTable[BottomTable >=1 ] <- 1
  
  GeneCol <- dplyr::select(table, c(gene.col))
  
  AllelesRaw <- base::cbind(GeneCol, topTable)
  base::names(AllelesRaw)[1] <- 'gene'
  
  Alleles <- stats::aggregate(. ~ gene, data=AllelesRaw, FUN=sum)
  
  SnpsCarrierRaw <- base::cbind(GeneCol, BottomTable)
  base::names(SnpsCarrierRaw)[1] <- 'gene'
  
  Snps <- stats::aggregate(. ~ gene, data=SnpsCarrierRaw, FUN=sum)
  
  Carrier <- Snps
  Carrier[Carrier >=1 ] <- 1
  
  if (flag == "Alleles"){
    return (Alleles)
  }
  else if (flag == "Snps"){
    return (Snps)}
  else if (flag == "Carrier"){
    return (Carrier)
  }

}

  
```


```{r}
library(tidyverse)
library(dplyr)

input <- tibble(CHROM=1, 
                POS=2:7, 
                ID = ".", 
                GENE=c("g1", "g1", "g2", "g3", "g4","g4"), 
                s1=c("0/1", "1/0", rep("0/0", 4)),
                s2=c("1/0", "0/0", "1/0", "0/0", "0/1", "./."),
                s3=c("0/0", "1/1", "0/0","0/0", "./.", "./."),
                s4=c(rep("0/0", 3), "0/1", "1/1", "0/1"))
```

```{r}
count_gene_mutations (input, 4, 5, "Carrier")
```





